File: leetcode/Binary_Tree_Level_Order_Traversal_II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> levelOrderBottom(TreeNode root) {
        ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
        Queue<TreeNode>[] q = (Queue<TreeNode>[])new Queue[2];
        q[0] = new LinkedList<TreeNode>();
        if (root != null) {
            q[0].offer(root);
        }
        for (int i = 0; ; ++i) {
            int a = i % 2;
            int b = 1 - a;
            if (q[a].isEmpty()) break;
            q[b] = new LinkedList<TreeNode>();
            ArrayList<Integer> r = new ArrayList<Integer>();
            while (!q[a].isEmpty()) {
                TreeNode t = q[a].poll();
                r.add(t.val);
                if (t.left != null) q[b].offer(t.left);
                if (t.right != null) q[b].offer(t.right);
            }
            ret.add(0, r);
        }
        return ret;
    }
}

File: leetcode/Binary_Tree_Level_Order_Traversal_II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int max_depth;
    void findMaxDepth(TreeNode *node, int depth) {
        if (node == NULL)
            return;
        if (max_depth < depth)
            max_depth = depth;
        findMaxDepth(node->left, depth + 1);
        findMaxDepth(node->right, depth + 1);
    }
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        max_depth = 0;
        findMaxDepth(root, 1);
        
        vector<vector<int> > traversal(max_depth);
        if (root == NULL)
            return traversal;
            
        queue<TreeNode*> node_queue;
        node_queue.push(root);
        queue<int> idx_queue;
        idx_queue.push(max_depth - 1);
        while (!node_queue.empty()) {
            TreeNode *node = node_queue.front();
            node_queue.pop();
            int idx = idx_queue.front();
            idx_queue.pop();
            traversal[idx].push_back(node->val);
            if (node->left != NULL) {
                node_queue.push(node->left);
                idx_queue.push(idx - 1);
            }
            if (node->right != NULL) {
                node_queue.push(node->right);
                idx_queue.push(idx - 1);
            }
        }
        return traversal;
    }
};
